numpyro/mcmc.py

```diff
@@ -0,0 +1,124 @@
+import math
+
+import jax
+import jax.numpy as np
+from jax import lax, partial, random
+from jax.flatten_util import ravel_pytree
+from jax.random import PRNGKey
+
+import numpyro.distributions as dist
+from numpyro.hmc_util import IntegratorState, find_reasonable_step_size, velocity_verlet, warmup_adapter
+from numpyro.util import laxtuple
+
+HMCState = laxtuple('HMCState', ['z', 'z_grad', 'potential_energy', 'num_steps',
+                                 'step_size', 'inverse_mass_matrix', 'rng'])
+
+
+def _get_num_steps(step_size, trajectory_length):
+    num_steps = np.array(trajectory_length / step_size, dtype=np.int32)
+    return np.where(num_steps < 1, np.array(1, dtype=np.int32), num_steps)
+
+
+def _sample_momentum(inverse_mass_matrix, pack_fn, rng):
+    if inverse_mass_matrix.ndim == 1:
+        return pack_fn(dist.norm(0., np.sqrt(np.reciprocal(inverse_mass_matrix))).rvs(random_state=rng))
+    elif inverse_mass_matrix.ndim == 2:
+        raise NotImplementedError
```

Comment:
I think that we can sample `epsilon ~ N(0, 1)` and instead of `inverse_mass_matrix`, we require an input "scale" here. We might have a MassMatrixState(inverse, sqrt) laxtuple and will update it at warmup_adapter method. The inverse one is used for kinetic_fn and sqrt is used for sampling momentum.

Reply:
That's a good way to do it.

Reply:
I would like to punt this to a follow-up PR, if that's okay.